WireDomain/Sources/WireDomain/Repositories/UpdateEventsRepository.swift

```diff
@@ -0,0 +1,122 @@
+    ///
+    /// Pending events are events that have been buffered by the server while
+    /// the self client has not had an active push channel.
+
+    func pullPendingEvents() async throws
+
+}
+
+final class UpdateEventsRepository: UpdateEventsRepositoryProtocol {
+
+    private let selfClientID: String
+    private let updateEventsAPI: any UpdateEventsAPI
+    private let updateEventDecryptor: any UpdateEventDecryptorProtocol
+    private let eventContext: NSManagedObjectContext
+    private let jsonEncoder = JSONEncoder()
+
+    private let lastEventIDRepository: any LastEventIDRepositoryInterface
+
+    init(
+        selfClientID: String,
+        updateEventsAPI: any UpdateEventsAPI,
+        updateEventDecryptor: any UpdateEventDecryptorProtocol,
+        eventContext: NSManagedObjectContext,
+        lastEventIDRepository: any LastEventIDRepositoryInterface
+    ) {
+        self.selfClientID = selfClientID
+        self.updateEventsAPI = updateEventsAPI
+        self.updateEventDecryptor = updateEventDecryptor
+        self.eventContext = eventContext
+        self.lastEventIDRepository = lastEventIDRepository
+    }
+
+    func pullPendingEvents() async throws {
+        // We want all events since this event.
+        guard let lastEventID = lastEventIDRepository.fetchLastEventID() else {
+            throw UpdateEventsRepositoryError.lastEventIDMissing
+        }
+
+        // We'll insert new events from this index.
+        var currentIndex = try await indexOfLastEventEnvelope() + 1
+
+        // Events are fetched in batches.
+        for try await envelopes in updateEventsAPI.getUpdateEvents(
+            selfClientID: selfClientID,
+            sinceEventID: lastEventID
+        ) {
+            for envelope in envelopes {
+                // We can only decrypt once so store the decrypted events for later retrieval.
+                var decryptedEnvelope = envelope
+                decryptedEnvelope.events = try await updateEventDecryptor.decryptEvents(in: envelope)
+
+                try await persistEventEnvelope(
+                    decryptedEnvelope,
+                    index: currentIndex
+                )
+
+                currentIndex += 1
+
+                if !envelope.isTransient {
+                    // Update the last event id so we don't refetch the same events.
+                    // Transient events aren't stored in the backend's event stream.
+                    lastEventIDRepository.storeLastEventID(envelope.id)
+                }
+            }
+        }
+    }
+
+    private func indexOfLastEventEnvelope() async throws -> Int64 {
+        try await eventContext.perform { [eventContext] in
+            let request = StoredUpdateEventEnvelope.sortedFetchRequest(asending: false)
+            request.fetchBatchSize = 1
+            let lastEnvelope = try eventContext.fetch(request).first
+            return lastEnvelope?.sortIndex ?? 0
```

Comment:
Is it right to return `0` instead of an error?

Reply:
if there is no last envelope, then it's not an error, we just start with a zero index. The important thing isn't where the index starts, but that it increases monotonically.